File: ch6/clip2/boj1213/Main.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String name = sc.next();
        int[] cnt = new int[26];
        for(int i = 0; i < name.length(); i++) {
            cnt[name.charAt(i) - 'A']++;
        }

        int odd = 0;
        int oddIdx = -1;
        for(int i = 0; i < 26; i++) {
            if(cnt[i] % 2 == 1) {
                odd++;
                oddIdx = i;
            }
        }
        if(odd > 1) {
            System.out.println("I'm Sorry Hansoo");
            return;
        }

        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < 26; i++) {
            for(int j = 0; j < cnt[i] / 2; j++) {
                sb.append((char)(i + 'A'));
            }
        }
        StringBuilder ans = new StringBuilder(sb);
        if(oddIdx != -1) {
            ans.append((char)(oddIdx + 'A'));
        }
        ans.append(sb.reverse());
        System.out.println(ans);
    }
}

File: ch6/clip2/boj1213/main.py
name = input()
cnt = [0] * 26

for char in name:
    cnt[ord(char) - ord('A')] += 1

odd = 0
odd_idx = -1

for i in range(26):
    if cnt[i] % 2 == 1:
        odd += 1
        odd_idx = i

if odd > 1:
    print("I'm Sorry Hansoo")
else:
    sb = []
    for i in range(26):
        sb.extend([chr(ord('A') + i)] * (cnt[i] // 2))

    ans = sb[:]
    if odd_idx != -1:
        ans.append(chr(ord('A') + odd_idx))
    
    ans.extend(reversed(sb))
    print(''.join(ans))
